Qt: Add Android packaging support

icota opened the pull request:
![bitcoin-qt](https://user-images.githubusercontent.com/762502/67396157-62f3d000-f5a7-11e9-8a6f-9425823fcd6c.gif)
This PR is the third and final piece of the basic Android support puzzle - it depends on https://github.com/bitcoin/bitcoin/pull/16110 and is related to https://github.com/bitcoin/bitcoin/pull/16883. It introduces an `android` directory under `qt` and a simple way to build an Android package of `bitcoin-qt`:

1. Build depends for Android as described in the [README](https://github.com/bitcoin/bitcoin/blob/master/depends/README.md)
2. Configure with one of the resulting prefixes
3. Run `make && make apk` in `src/qt`

The resulting APK files will be in `android/build/outputs/apk`. You can install them manually or with [adb](https://developer.android.com/studio/command-line/adb). One can also open the `android` directory in Android Studio for that integrated development and debugging experience. `BitcoinQtActivity` is your starting point.

Under the hood makefile `apk` target:

1. Renames the `bitcoin-qt` binary to `libbitcoin-qt.so` and copies it over to a folder under `android/libs` depending on which prefix and corresponding [ABI](https://developer.android.com/ndk/guides/abis.html#sa) `bitcoin-qt` was built for
2. Takes `libc++_shared.so` from the Android NDK and puts in the same place. It [must be included](https://developer.android.com/ndk/guides/cpp-support) in the APK
3. Extracts Qt for Android Java support files from the `qtbase` archive in `depends/sources` to `android/src`

There is also just a tiny bit of `ifdef`'d code to make the Qt Widgets menus usable. It's not pretty but it works and is a stepping stone towards https://github.com/bitcoin/bitcoin/pull/16883.

laanwj commented:
Concept ACK, nice !

icota commented:
Rebased, squashed, added instructions under `doc` and updated the first comment to reflect that #16110 has been merged.

MarcoFalke commented on doc/package-android.md:
```diff
@@ -0,0 +1,12 @@
+Android package guide
```
Should this file be linked to from the doc/Readme?

Also, should it be named build-android, similar to build-unix and build-bsd?

icota replied:
I thought so but other build files seem to be about building _on_ UNIX, BSD, etc... I'm open to suggestions.

MarcoFalke replied:
Oh, this is for cross-compilation? Well, we have build-win

icota replied:
You are right. I will change to build-android and make it a bit more generic (as in not just about packaging)

luke-jr commented:
> This PR is the third and final piece of the basic Android support puzzle

IMO final would be a gitian descriptor ;)

adamjonas commented:
@icota Friendly ping for a rebase.

icota commented:
Thank you @adamjonas. Rebased.

MarcoFalke commented:
@icota Could you pull in https://github.com/MarcoFalke/bitcoin-core/commit/a0b441baf9dbb3ce7e332b0a8e0d4d278e73f611 to shave off a few kB off of the compressed images?

icota commented:
Absolutely. Thanks @MarcoFalke

MarcoFalke commented:
Oh, you'll have to squash it. Otherwise git will keep track of both versions, thus doubling the size.

icota commented:
Done.

BlockMechanic commented:
Awesome, will test this out.

jonasschnelli commented:
utACK a818b2c082e257eb67403479df61f1325231c947

icota commented:
Yep it needs the SDK. Travis seems to [support this](https://docs.travis-ci.com/user/languages/android/).

MarcoFalke commented:
Unfortunately (or luckily) we can't use travis' packages. We use our own ci setup in `./ci`.

The sdkmanager is a zip file that can be downloaded, though. And then a one-liner like

```
$ANDROID_HOME/tools/bin/sdkmanager --sdk_root=$ANDROID_HOME "platform-tools" "platforms;android-28" "ndk-bundle"
``` 

should do the rest.

jovi55 commented:
کی شروع به کارمیکند

jonasschnelli commented:
I'd like to give this a try in the main branch (even without CI support). If it gets unmaintained, a revert is likely.
Anyone with Android experience willing to test this @Sjors, @hebasto, @promag, @greenaddress?

MarcoFalke commented:
To clarify. I did try to build this, but the linker step of bitcoind failed, which is why I asked for a `./ci` file or at least some indication that anyone has tried compiling this current version of the pull merged into current master. I am generally not a fan of merging code and build system changes that may or may not compile.

I don't have android experience, so maybe I did something wrong.

icota commented:
@MarcoFalke On Linux you might have to pull https://github.com/bitcoin/bitcoin/pull/17396. Specifically I think the linker error is due to an autoconf issue. A fix has since been [merged upstream](https://github.com/autoconf-archive/autoconf-archive/pull/211).

BlockMechanic commented:
> To clarify. I did try to build this, but the linker step of bitcoind failed, which is why I asked for a `./ci` file or at least some indication that anyone has tried compiling this current version of the pull merged into current master. I am generally not a fan of merging code and build system changes that may or may not compile.
> 
> I don't have android experience, so maybe I did something wrong.

I found some issues in the alqo_qt.m4 file. Mine now compiles, now trying to produce a apk. 

@icota which version of grade is best ? i see you were using 3.2, but the latest is 6.x. Should we have a recommended version ?

icota commented:
> @icota which version of grade is best ? i see you were using 3.2, but the latest is 6.x. Should we have a recommended version ?
I don't think so, we are not using much build system wizardry here anyway... Unless you have a specific reason to bump Gradle, I'd keep it at 3.2

MarcoFalke commented:
Needs rebase?

icota commented:
Rebased. I'll look into adding this to CI.

icota commented:
@MarcoFalke Can you take a look at the CI changes? The general idea is to have the Android CLI tools fetch everything needed to build and having done that we 'iterate' over the 4 supported ABIs so the final APK built will have contained all the binaries.

I haven't actually tested this to completion (it's going to take some time on my machine) but would like some feedback on the direction.

MarcoFalke commented on ci/test/06_script_a.sh:
```diff
@@ -20,6 +20,26 @@ END_FOLD
+  DOCKER_EXEC make $MAKEJOBS $GOAL || ( echo "Build failure. Verbose build follows." && DOCKER_EXEC make $GOAL V=1 ; false )
+  DOCKER_EXEC make -C src/qt apk
+
+  DOCKER_EXEC "${BASE_ROOT_DIR}/configure" --prefix=$DEPENDS_DIR/x86_64-linux-android || ( (DOCKER_EXEC cat config.log) && false)
```
It is probably enough to keep only the `aarch64-linux-android` build.

icota replied:
Agreed. Okay to just comment the others out in case somebody wants them?

MarcoFalke replied:
I think it would be easy to re-create them by copy-paste and adjusting the name, etc.

maflcko reviewed (commented):
ci looks ok. You can also keep the ci changes in one separate commit, so if it turns out to not work, the commit can simply be dropped/reverted on its own

icota committed:
Qt: add Android packaging support

Introduce an android directory under qt and allow one to package bitcoin-qt for Android by running make apk.
Add bitcoin-qt Android build instructions.

icota committed:
CI: add Android APK build script

icota commented:
To build with CI run:
`FILE_ENV="./ci/test/00_setup_env_android.sh" ./ci/test_run_all.sh`

To install on your (ARM64) device:
`docker cp ci_android:$PWD/src/qt/android/build/outputs/apk/debug/android-debug.apk .`
`adb install android-debug.apk`

MarcoFalke commented:
Could add the ci file to travis.yml or cirrus.yml? If it takes 30-40 minutes, then travis should be sufficient. If it takes longer, then cirrus might be better. If it takes longer than 2h, maybe leave it out for now.

icota commented:
@MarcoFalke It's about 30 mins to build just for ARM64 so I added it to travis.

luke-jr commented on src/qt/bitcoin.cpp:
```diff
@@ -460,6 +460,13 @@ int GuiMain(int argc, char* argv[])
     // Generate high-dpi pixmaps
     QApplication::setAttribute(Qt::AA_UseHighDpiPixmaps);
+
+#if defined(QT_QPA_PLATFORM_ANDROID)
+    QApplication::setAttribute(Qt::AA_DontUseNativeMenuBar);
+    QApplication::setAttribute(Qt::AA_DontCreateNativeWidgetSiblings);
+    QApplication::setAttribute(Qt::AA_DontUseNativeDialogs);
+#endif
```
Why not??

icota replied:
![image](https://user-images.githubusercontent.com/762502/97157083-eaa28a80-1777-11eb-8c08-d73e2291bd6b.png)
Native menus (or Qt's Android native implementation of QMenu) don't support submenus. When you tap on any of the four options nothing happens.

The other two attributes are there because native dialogs are way too simple to render our modals and to make sure the dialogs don't create another window (which is only supported on ChromeOS).

decryp2kanon commented:
Concept ACK

MarcoFalke commented on .travis.yml:
```diff
@@ -94,6 +94,11 @@ jobs:
+    - stage: test
+      name: 'ARM64 Android APK  [GOAL: build]'
+      env: >-
+        FILE_ENV="./ci/test/00_setup_env_android.sh"
```
We no longer use travis, so you'll have to move those added lines over to cirrus

icota replied:
Done

icota committed:
CI: add Android APK build to cirrus

icota commented:
I'm counting four concept ACKs. For anyone looking to test it's as easy as:
> To build with CI run:
> `FILE_ENV="./ci/test/00_setup_env_android.sh" ./ci/test_run_all.sh`
> 
> To install on your (ARM64) device:
> `docker cp ci_android:$PWD/src/qt/android/build/outputs/apk/debug/android-debug.apk .`
> `adb install android-debug.apk`

MarcoFalke commented:
This hasn't been merged because we are still in feature freeze for 0.21, but this should be good to go into 22.0

maflcko reviewed (commented):
Tested compilation locally (via ci script) in podman:

```
Options used to compile and link:
  boost process = no
  multiprocess  = no
  with wallet   = yes
    with sqlite = yes
  with gui / qt = yes
    with qr     = yes
  with zmq      = yes
  with test     = yes
    with fuzz   = no
  with bench    = yes
  with upnp     = yes
  use asm       = yes
  sanitizers    = 
  debug enabled = no
  gprof enabled = no
  werror        = no

  target os     = android
  build os      = linux-gnu

  CC            = /bitcoin/depends/SDKs/android/ndk/21.1.6352462/toolchains/llvm/prebuilt/linux-x86_64/bin//aarch64-linux-android28-clang
  CFLAGS        = -pthread -g -O2
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I/bitcoin/depends/aarch64-linux-android/share/../include/  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = /bitcoin/depends/SDKs/android/ndk/21.1.6352462/toolchains/llvm/prebuilt/linux-x86_64/bin//aarch64-linux-android28-clang++ -std=c++11
  CXXFLAGS      =   -Wstack-protector -fstack-protector-all  -Wall -Wextra -Wgnu -Wformat -Wformat-security -Wvla -Wshadow-field -Wswitch -Wthread-safety -Wrange-loop-analysis -Wredundant-decls -Wunused-variable -Wdate-time -Wconditional-uninitialized -Wsign-compare -Woverloaded-virtual -Wunreachable-code-loop-increment  -Wno-unused-parameter -Wno-self-assign -Wno-unused-local-typedef -Wno-deprecated-register -Wno-implicit-fallthrough   -g -O2
  LDFLAGS       =   -Wl,-z,relro -Wl,-z,now -pie  -L/bitcoin/depends/aarch64-linux-android/share/../lib 
  ARFLAGS       = cr

MarcoFalke commented:
ACK 11be2d282965759edd87284d111f067168d937c2

BlockMechanic commented:
With static depends bui;ds now a thing, i think this one just moved up the list :) Will run tests later today.

prayank23 commented:
Concept ACK

This looks interesting. Not sure how did I miss this. Also interested to know difference and trade-offs in this solution compared to [ABCore](https://github.com/greenaddress/abcore) or [NayutaCore](https://github.com/nayutaco/NayutaCore) and if more projects exist to use bitcoin core on Android.

laanwj commented:
> This looks interesting. Not sure how did I miss this.

Haha I wonder this all the time. So much going on.

Looks like the new APK build is failing on the Qt compile. Maybe the intermediate Qt version bump got in the way?
```
ERROR: Feature 'precompile_header' was enabled, but the pre-condition 'tests.precompile_header' failed.
funcs.mk:273: recipe for target '/tmp/cirrus-ci-build/depends/work/build/aarch64-linux-android/qt/5.12.10-3ab96e3dc5c/./.stamp_configured' failed
make: *** [/tmp/cirrus-ci-build/depends/work/build/aarch64-linux-android/qt/5.12.10-3ab96e3dc5c/./.stamp_configured] Error 3
make: Leaving directory '/tmp/cirrus-ci-build/depends'
```

icota committed:
depends: disable Qt Vulkan support on Android

icota committed:
depends: fix Qt precompiled headers bug

Fixed in 5.14, see QTBUG-85214

icota commented:
@laanwj Thanks for the heads up. I rebased on top of 5.12 and tested on an Android 11 device.

@prayank23 I can't comment on other approaches. This PR aims to bring nothing less (and nothing more) than `bitcoin-qt` running on Android. https://github.com/bitcoin/bitcoin/pull/16883 is an attempt to make the GUI a bit more modern and usable but it's parked for now as I can't find the time to work on it.

BTW everyone check out the testnet IBD:
![Screenshot_20210321-161502](https://user-images.githubusercontent.com/762502/111911965-e5681400-8a67-11eb-9128-1dd5315a01fb.jpg)
That seems quite fast, no? And the phone's not even warm.... This PR might make practical sense after all. :sunglasses:

MarcoFalke commented:
cr ACK 246774e264

laanwj commented:
Code review ACK 246774e26459cb3652e308880abdd140e8e9d204

prayank23 commented:
> #16883 is an attempt to make the GUI a bit more modern and usable but it's parked for now as I can't find the time to work on it.

@icota Thanks for working on this. Its merged :) #16883 looks interesting except one comment from @promag , my younger brother is also good with Android development but Kotlin. Will share things with him after testing.

ryanofsky commented:
Seeing a CI failure in ARM64 Android APK [bionic] build that might be related to this:

https://cirrus-ci.com/task/5511109198217216?command=ci#L4420

```
make[1]: Entering directory '/tmp/cirrus-ci-build/src'
tar: --exclude=*/*: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
tar: --exclude=*/*: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
tar: --exclude=*/*: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
mkdir -p qt/android/libs/arm64-v8a
cp /tmp/cirrus-ci-build/depends/SDKs/android/ndk/21.1.6352462/toolchains/llvm/prebuilt/linux-x86_64/bin//../sysroot/usr/lib/aarch64-linux-android/libc++_shared.so qt/android/libs/arm64-v8a
tar xf  -C qt/android/src/ src/android/jar/src --strip-components=5
tar: -C: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
Makefile:15443: recipe for target 'bitcoin_qt_apk' failed
```

icota commented:
@ryanofsky Yes, and thisline is the culprit:
https://github.com/bitcoin/bitcoin/blob/9217f9fe7351d9c36803456f88ae561e23f17ac9/src/Makefile.qt.include#L382

Is Cirrus perhaps wiping `depends/sources` from time to time?

achow101 commented:
> Is Cirrus perhaps wiping `depends/sources` from time to time?

I'm able to reproduce this error locally so I don't think it is a cirrus issue.

icota commented:
I can't reproduce locally (unless I manually delete `sources`) but I'm seeing a pattern on the latest runs:

fails when run last:
https://github.com/bitcoin/bitcoin/pull/21392/checks
https://github.com/bitcoin/bitcoin/pull/16545/checks
https://github.com/bitcoin/bitcoin/pull/19460/checks

succeds otherwise:
https://github.com/bitcoin/bitcoin/pull/21422/checks
https://github.com/bitcoin/bitcoin/pull/21526/checks
https://github.com/bitcoin/bitcoin/pull/17227/checks

I'm drawing a blank here and I am also fine with disabling the CI APK build for the time being.

icota commented:
Trying to fix the issue in https://github.com/bitcoin/bitcoin/pull/21537

icota commented:
Fixed by @MarcoFalke in https://github.com/bitcoin/bitcoin/pull/21541

hebasto commented on src/Makefile.qt.include:
```diff
@@ -378,6 +378,20 @@ bitcoin_qt_clean: FORCE
 
 bitcoin_qt : qt/bitcoin-qt$(EXEEXT)
 
+APK_LIB_DIR = qt/android/libs/$(ANDROID_ARCH)
```
Shouldn't the content of the `qt/android` directory be a subject of the cleanup in `make clean` or `make distclean`?

icota replied:
It should :+1:

hebasto commented on src/qt/android/src/org/bitcoincore/qt/BitcoinQtActivity.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public void onCreate(Bundle savedInstanceState)
+    {
+        final File bitcoinDir = new File(getFilesDir().getAbsolutePath() + "/.bitcoin");
+        if (!bitcoinDir.exists()) {
+            bitcoinDir.mkdir();
+        }
```
This code prevents the Intro dialog from showing at the first startup. Is it intended? If yes, why?

icota replied:
TBH I don't remember why I did this. Did you try without it? Maybe the dialog was showing incorrectly?

I didn't at the time but I realise now that this makes using external storage impossible. Possibly related to https://github.com/bitcoin-core/gui/pull/496

hebasto commented:
After bumping Qt to 5.15.2 (#23489) this stuff seems required to be updated to work properly.
Massive changes were done in Qt 5.14 (e.g., https://github.com/qt/qtbase/commit/5bb178c479a247720fbc3fbb7f06a32b725193ac).

@icota Could you have a look into this please?

hebasto commented:
> After bumping Qt to 5.15.2 (#23489) this stuff seems required to be updated to work properly. Massive changes were done in Qt 5.14 (e.g., [qt/qtbase@5bb178c](https://github.com/qt/qtbase/commit/5bb178c479a247720fbc3fbb7f06a32b725193ac)).
> 
> @icota Could you have a look into this please?

https://www.kdab.com/qt-for-android-better-than-ever-before/ seems related.

icota commented:
Thanks for letting me know @hebasto. Attempting to fix the issues in https://github.com/bitcoin-core/gui/pull/504
